src/main/java/uk/gov/justice/digital/hmpps/oauth2server/service/MfaService.kt

```
@@ -49,24 +51,37 @@ open class MfaService(@Value("\${application.authentication.mfa.whitelist}") whi
     val token = tokenService.createToken(TokenType.MFA, username)
     val code = tokenService.createToken(TokenType.MFA_CODE, username)
 
-    val firstName = userService.findMasterUserPersonDetails(username).map { it.firstName }.orElseThrow()
-
-    notificationClient.sendEmail(mfaTemplateId, user.email, mapOf("firstName" to firstName, "code" to code), null)
+    emailCode(user, code)
 
     return Pair(token, code)
   }
 
-  open fun validateMfaCode(code: String?): Optional<String> {
-    if (code.isNullOrBlank()) return Optional.of("missingcode");
-    // 1. look up mfa code
-    // 2. fail if expired or invalid - incorrect password attempts?
-    // 3. return empty if okay
-    return Optional.empty()
+  @Transactional(transactionManager = "authTransactionManager")
+  open fun validateAndRemoveMfaCode(token: String, code: String?): Optional<String> {
```

Any good reason not make the return String nullable instead of Optional?

---

I was chatting to Jamie about this yesterday.  Not really happy with Kotlin having to return`null` instead of `Optional`, since we use `Optional` instead of `null` in java...

---

Can have another look next week!

---

I'm not that keen about the `Optional` stuff in the first place (even though I added it all).  With hindsight would be better to just throw an exception and then catch it, would be clearer what is going on as in some cases the optional empty is success, whereas in other cases it is failure...